### PR TITLE
begin updating webhook documentation

### PR DESCRIPTION
**Pull request by lavalamp**
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> For 1.9 Features: set Milestone to `1.9` and Base Branch to `release-1.9`
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> NOTE: Please check the “Allow edits from maintainers” box (see image below) to
> [allow reviewers to fix problems](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) on your patch and speed up the review process.
>
> Please delete this note before submitting the pull request.
>
> NOTE: After opening the PR, please *un-check and re-check* the "Allow edits from maintainers" box. This is a temporary workaround to address a known issue with GitHub.

Allow edits from maintainers checkbox

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/website/6368)

### CODE REVIEW

**Comment by lavalamp** [comment]
I don't think this is done yet but I want to get something out there.

File: docs/admin/admission-controllers.md [hunk 1]
```diff
@@ -16,16 +16,27 @@ title: Using Admission Controllers
 
 An admission control plug-in is a piece of code that intercepts requests to the Kubernetes
 API server prior to persistence of the object, but after the request is authenticated
-and authorized.  The plug-in code is in the API server process
-and must be compiled into the binary in order to be used at this time.
+and authorized.  The plug-ins discussed in this document are compiled into the
+API server process and may only be configured by the cluster administrator. As of 1.9, [user provided
+webhooks](/docs/admin/extensible-admission-controllers.md#external-admission-webhooks)
+are in beta; these are dynamically configurable.
```
Comment (erictune) [inline docs/admin/admission-controllers.md]: _runtime configurable_ is more specific than _dynamically configurable_?

File: docs/admin/admission-controllers.md [hunk 2]
```diff
@@ -16,16 +16,27 @@ title: Using Admission Controllers
 
 An admission control plug-in is a piece of code that intercepts requests to the Kubernetes
 API server prior to persistence of the object, but after the request is authenticated
-and authorized.  The plug-in code is in the API server process
-and must be compiled into the binary in order to be used at this time.
+and authorized.  The plug-ins discussed in this document are compiled into the
```
Comment (erictune) [inline docs/admin/admission-controllers.md]: Let's expunge the term plug-ins.
We don't want to accept more, and we don't want to encourage forking, so why use a term which invites plugging things in?
Maybe "modules".

Comment (erictune) [reply docs/admin/admission-controllers.md]: Or just "admission controller"?

File: docs/admin/admission-controllers.md [hunk 3]
```diff
@@ -16,16 +16,27 @@ title: Using Admission Controllers
...
+API server process and may only be configured by the cluster administrator. As of 1.9, [user provided
+webhooks](/docs/admin/extensible-admission-controllers.md#external-admission-webhooks)
+are in beta; these are dynamically configurable.
 
-Each admission control plug-in is run in sequence before a request is accepted into the cluster.  If
-any of the plug-ins in the sequence reject the request, the entire request is rejected immediately
-and an error is returned to the end-user.
+Admission control plug-ins may be "validating", "mutating", or both. Mutating
+plug-ins may modify the objects they admit; validating plug-ins may not.
```
Comment (erictune) [inline docs/admin/admission-controllers.md]: I'm a reader unfamiliar with this: I can't tell if the word "plug-ins"  includes or excludes "user provided webhooks"

File: docs/admin/admission-controllers.md [hunk 4]
```diff
@@ -101,14 +112,16 @@ If your cluster supports containers that run with escalated privileges, and you
 restrict the ability of end-users to exec commands in those containers, we strongly encourage
 enabling this plug-in.
 
-### GenericAdmissionWebhook (alpha)
+### GenericAdmissionWebhook (alpha in 1.8; beta in 1.9)
```
Comment (erictune) [inline docs/admin/admission-controllers.md]: We now have the following terms that appear to mean the same thing:
- extensible-admission-controllers
- external-admission-webhooks
- Dynamic Admission Control
- External Admission Webhooks

Comment (erictune) [inline docs/admin/admission-controllers.md]: Why did this get called Generic, when we use the term Validating elsewhere.  Wasn't naming settled in the design phase?

Comment (lavalamp) [reply docs/admin/admission-controllers.md]: I intend to fix the names before this is finalized.

Comment (lavalamp) [reply docs/admin/admission-controllers.md]: Chao is already on it: https://github.com/kubernetes/kubernetes/pull/55988

Comment (lavalamp) [reply docs/admin/admission-controllers.md]: It is Validating. Will fix this when https://github.com/kubernetes/kubernetes/pull/55988 merges.

File: docs/admin/admission-controllers.md [hunk 5]
```diff
@@ -426,6 +450,11 @@ We strongly recommend using this plug-in if you intend to make use of Kubernetes
 ## Is there a recommended set of plug-ins to use?
 
 Yes.
+For Kubernetes >= 1.9.0, we strongly recommend running the following set of admission control plug-ins (order matters):
+
+```shell
+--admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,GenericAdmissionWebhook,ResourceQuota,DefaultTolerationSeconds,MutatingAdmissionWebhook
```
Comment (deads2k) [inline docs/admin/admission-controllers.md]: @kubernetes/sig-auth-api-reviews who's ready to add PSP to the list?

Comment (deads2k) [reply docs/admin/admission-controllers.md]: Others to add:
 1. OwnerReferencesPermissionEnforcement
 2. NodeRestriction
 3. PodNodeSelector
 4. PodTolerationRestriction - @kubernetes/sig-scheduling-api-reviews the annotation have been alpha for a while, but the taints are not alpha.  Ready for this?

File: docs/admin/admission-controllers.md (hunk 1 above)
Comment (sttts) [reply docs/admin/admission-controllers.md]: Confusing to talk about plugins and then suddenly about webhooks. Question a read might have: is a webhook a plugin?

File: docs/admin/admission-controllers.md [hunk 6]
```diff
@@ -16,16 +16,27 @@ title: Using Admission Controllers
...
+Admission control plug-ins may be "validating", "mutating", or both. Mutating
+plug-ins may modify the objects they admit; validating plug-ins may not.
 
-Admission control plug-ins may mutate the incoming object in some cases to apply system configured
-defaults.  In addition, admission control plug-ins may mutate related resources as part of request
-processing to do things like increment quota usage.
+The admission control process proceeds in two phases. In the first phase,
+mutating admission control plug-ins are run. In the second phase, validating
+admission control plug-ins are run. Note again that some of the plug-ins are
+both. In both phases, the plug-ins are run in the order specified by the
```
Comment (sttts) [inline docs/admin/admission-controllers.md]: only in the mutating phase the order matters and is defined.

File: docs/admin/admission-controllers.md (hunk 4 above)
Comment (sttts) [reply docs/admin/admission-controllers.md]: It's called ValidatingAdmissionWebhook

File: docs/admin/admission-controllers.md (hunk 5 above)
Comment (sttts) [reply docs/admin/admission-controllers.md]: it's called ValidatingAdmissionWebhook

File: docs/admin/admission-controllers.md [hunk 7]
```diff
@@ -256,6 +269,17 @@ applies a 0.1 CPU requirement to all Pods in the `default` namespace.
...
+
+This plug-in calls any mutating webhooks which match the request. Matching
+webhooks are called in serial; each one may modify the object if it desires.
+
+This plugin (as implied by the name) only runs in the mutating phase.
+
+If a webhook called by this has side effects (e.g., decrementing quota) it
+*must* have a reconcilation system, as it is not guaranteed that subsequent
+webhooks or validating plugins will permit the request to finish.
+
```
Comment (erictune) [inline docs/admin/admission-controllers.md]: Add the following text:

Use caution when authoring and installing mutating webhooks.  
*  Users may be confused when the objects they try to create are different from what they get back. 
  * Setting originally unset fields is less confusing that overwriting fields set in the request.  Avoid doing the latter.  
* Built in control loops may break when the objects they try to create are different when read back.
  * Setting originally unset fields is less likely to cause problems than overwriting fields set in the original request.  Avoid doing the latter.
* This is a beta feature.  Future versions of Kubernetes may restrict the types of mutating webhooks can make.
* Future changes to control loops for built-in resources or third-party resources may break webhooks that work well today.  Even when the webhook installation API is field, not all possible webhook behaviors will be guaranteed to be supported indefinitely.

File: docs/admin/admission-controllers.md [hunk 8]
```diff
@@ -101,14 +112,16 @@ If your cluster supports containers that run with escalated privileges, and you
...
+fails. This plugin only runs in the validation phase; the webhooks it calls may not
+mutate the object, as opposed to the webhooks called by the `MutatingAdmissionWebhook` plugin.
 
-This plug-in is related to the [Dynamic Admission Control](/docs/admin/extensible-admission-controllers)
-introduced in v1.7.
-The plug-in calls the webhooks configured via `ExternalAdmissionHookConfiguration`,
-and only admits the operation if all the webhooks admit it.
-Currently, the plug-in always fails open.
-In other words, it ignores the failed calls to a webhook.
+If a webhook called by this has side effects (e.g., decrementing quota) it
```
Comment (zacharysarah) [inline docs/admin/admission-controllers.md]: > If a webhook called by this has side effects (for example, decrementing quota), it

File: docs/admin/admission-controllers.md (hunk 7 above)
Comment (zacharysarah) [reply docs/admin/admission-controllers.md]: Some tweaks to @erictune's additions:
```
Use caution when authoring and installing mutating webhooks.
* This is a beta feature. Future versions of Kubernetes may restrict the types of mutations webhooks can make.
* Users may be confused when the objects they try to create are different from what they receive. 
* Avoid overwriting fields. Setting originally unset fields is less confusing than overwriting fields set in the request.  
* Built in control loops may break when the objects they try to create are different when read back.
* Future changes to control loops for built-in resources or third-party resources may break webhooks that work well today. Even when the webhook installation API is a field, not all possible webhook behaviors will be guaranteed to be supported indefinitely.
```
So:
> Use caution when authoring and installing mutating webhooks.
> * Users may be confused when the objects they try to create are different from what they receive. 
> * Avoid overwriting fields. Setting originally unset fields is less confusing than overwriting fields set in the request.  
> * Built in control loops may break when the objects they try to create are different when read back.
> * This is a beta feature. Future versions of Kubernetes may restrict the types of mutations webhooks can make.
> * Future changes to control loops for built-in resources or third-party resources may break webhooks that work well today. Even when the webhook installation API is a field, not all possible webhook behaviors will be guaranteed to be supported indefinitely.

File: docs/admin/admission-controllers.md [hunk 9]
```diff
@@ -426,6 +450,11 @@ We strongly recommend using this plug-in if you intend to make use of Kubernetes
 ## Is there a recommended set of plug-ins to use?
 
 Yes.
+For Kubernetes >= 1.9.0, we strongly recommend running the following set of admission control plug-ins (order matters):
```
Comment (zacharysarah) [inline docs/admin/admission-controllers.md]:
```
For Kubernetes >= 1.9.0, we strongly recommend running the following set of admission control plug-ins in the following order:
```
So:
> For Kubernetes >= 1.9.0, we strongly recommend running the following set of admission control plug-ins in the following order:

File: docs/admin/admission-controllers.md (hunk 5 above)
Comment (lavalamp) [reply docs/admin/admission-controllers.md]: Let's add those in a followup PR.

**Comment by lavalamp** [comment]
PTAL

**Comment by lavalamp** [comment]
The find-and-replaces are going to make rebasing this a huge pain, so the faster we can commit this, the better. Happy to make further changes in follow-ups if this meets some minimum bar.

**Comment by lavalamp** [comment]
I think there's some comments I didn't see that github hid after one of my edits--will try to double check later.

File: docs/admin/admission-controllers.md [hunk 10]
```diff
@@ -260,88 +269,121 @@ In any case, the annotations are provided by the user and are not validated by K
...
+This admission controller (as implied by the name) only runs in the mutating phase.
+
+If a webhook called by this has side effects (for example, decrementing quota) it
+*must* have a reconcilation system, as it is not guaranteed that subsequent
+webhooks or validating admission controllers will permit the request to finish.
+
+If you disable the MutatingAdmissionWebhook, you must also disable the
+`MutatingWebhookConfiguration` object in the `admissionregistration/v1beta1`
+group/version via the `--runtime-config` flag (both are on by default in
+versions >= 1.9).
```
Comment (caesarxuchao) [inline docs/admin/admission-controllers.md]: I would say it reversely: if you disable the API, you must disable the admission controller.

File: docs/admin/admission-controllers.md [hunk 11]
```diff
@@ -260,88 +269,121 @@ In any case, the annotations are provided by the user and are not validated by K
...
+
+#### Use caution when authoring and installing mutating webhooks
+
+ * Users may be confused when the objects they try to create are different from
+   what they get back.
+ * Setting originally unset fields is less confusing that overwriting fields set in
```
Comment (caesarxuchao) [inline docs/admin/admission-controllers.md]: Duplicate

File: docs/admin/admission-controllers.md [hunk 12]
```diff
@@ -440,47 +482,68 @@ The priority admission controller uses the `priorityClassName` field and populat
...
+mutate the object, as opposed to the webhooks called by the `MutatingAdmissionWebhook` admission controller.
+
+If a webhook called by this has side effects (for example, decrementing quota) it
+*must* have a reconcilation system, as it is not guaranteed that subsequent
+webhooks or other validating admission controllers will permit the request to finish.
 
+If you disable the ValidatingAdmissionWebhook, you must also disable the
+`ValidatingWebhookConfiguration` object in the `admissionregistration/v1beta1`
+group/version via the `--runtime-config` flag (both are on by default in
+versions >= 1.9).
```
Comment (caesarxuchao) [inline docs/admin/admission-controllers.md]: It's the reverse.

File: docs/admin/admission-controllers.md [hunk 13]
```diff
@@ -440,47 +482,68 @@ The priority admission controller uses the `priorityClassName` field and populat
...
-## Is there a recommended set of plug-ins to use?
+
+## Is there a recommended set of admission controllers to use?
 
 Yes.
-For Kubernetes >= 1.6.0, we strongly recommend running the following set of admission control plug-ins (order matters):
+For Kubernetes >= 1.9.0, we strongly recommend running the following set of admission controllers (order matters):
+
+```shell
+--admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,ValidatingAdmissionWebhook,ResourceQuota,DefaultTolerationSeconds,MutatingAdmissionWebhook
```
Comment (caesarxuchao) [inline docs/admin/admission-controllers.md]: Leaving `MutatingAdmissionWebhook` at last might be confusing.

**Comment by lavalamp** [comment]
Thanks for the comments, I will try to address them today.

**Comment by zacharysarah** [comment]
@lavalamp 👋 Sorry, this PR was merged prematurely. Please open a new PR and cherry-pick the commits, then continue working on the feedback from this PR in the new one.

/cc @tengqm

**Comment by lavalamp** [comment]
sigh :(

**Comment by zacharysarah** [comment]
@lavalamp Sorry. 😞

**Comment by zacharysarah** [comment]
@lavalamp 👋 Just a heads up that docs need to be complete by this Friday, 12/8. I'm happy to help with whatever I can; please let me know.

**Comment by lavalamp** [comment]
You could un-revert this PR? ;) ;)

More seriously I have three KubeCon sessions tomorrow and I am traveling
Friday.

What happens if it is not done on Friday?

**Comment by zacharysarah** [comment]
> You could un-revert this PR? ;) ;)

giphy 1

> More seriously I have three KubeCon sessions tomorrow and I am traveling Friday.

That's real.

> What happens if it is not done on Friday?

Can you have it done by the end of Monday?

It's already in pretty good shape, so I can arrange to review your PR last before we freeze and merge.

**Note:** After opening the new PR, be sure to **un-check and re-check** the box allowing edits from maintainers. (It's a workaround for a broken GitHub UI component.) This will shorten the amount of back and forth considerably.

Does that work?

**Comment by lavalamp** [comment]
Yeah, my plan was to work on it Monday, although I just looked at my
calendar and, yikes. I will do my best to get something presentable
sometime monday.

File: docs/admin/admission-controllers.md [hunk 14]
```diff
@@ -14,44 +14,62 @@ title: Using Admission Controllers
...
+admission controllers are run. Note again that some of the controllers are
+both. In both phases, the controllers are run in the order specified by the
+`--admission-control` flag of `kube-apiserver`.
+
+If any of the controllers in either phase reject the request, the entire
+request is rejected immediately and an error is returned to the end-user.
+
+Finally, in addition to sometimes mutating the object in question, admission
+controllers may sometimes have side effects, that is, mutate related
+resources as part of request processing. Incrementing quota usage is the
```
[the remark this answers is not in the document]
Comment (lavalamp) [reply docs/admin/admission-controllers.md]: This isn't talking about our specific quota mechanism, I think that's not a relevant detail.

File: docs/admin/admission-controllers.md [hunk 15]
```diff
@@ -260,88 +269,121 @@ In any case, the annotations are provided by the user and are not validated by K
...
 See the [limitRange design doc](https://git.k8s.io/community/contributors/design-proposals/resource-management/admission_control_limit_range.md) and the [example of Limit Range](/docs/tasks/configure-pod-container/limit-range/) for more details.
 
+### MutatingAdmissionWebhook (beta in 1.9)
+
+This admission controller calls any mutating webhooks which match the request. Matching
+webhooks are called in serial; each one may modify the object if it desires.
+
+This admission controller (as implied by the name) only runs in the mutating phase.
+
+If a webhook called by this has side effects (for example, decrementing quota) it
```
[the remark this answers is not in the document]
Comment (lavalamp) [reply docs/admin/admission-controllers.md]: This isn't talking about our specific quota mechanism, I think that's not a relevant detail.

File: docs/admin/admission-controllers.md (hunk 10 above)
Comment (lavalamp) [reply docs/admin/admission-controllers.md]: Having the admission controller but not the api on will just cause error messages in the logs. The other way around people might add webhooks and be confused about why they aren't running, and there'd be no error messages anywhere.

File: docs/admin/admission-controllers.md (hunk 11 above)
Comment (lavalamp) [reply docs/admin/admission-controllers.md]: fixed in new PR as soon as I push

File: docs/admin/admission-controllers.md (hunk 12 above)
Comment (lavalamp) [reply docs/admin/admission-controllers.md]: same response.

File: docs/admin/admission-controllers.md (hunk 13 above)
Comment (lavalamp) [reply docs/admin/admission-controllers.md]: Added more explanation instead. It's more important to explain how the flag works IMO.
